lua/luasnip/util/environ.lua

```diff
@@ -1,190 +1,139 @@
-local function _fill_eager_vars(env, pos)
-	env.TM_CURRENT_LINE =
-		vim.api.nvim_buf_get_lines(0, pos[1], pos[1] + 1, false)[1]
-	env.TM_CURRENT_WORD = util.word_under_cursor(pos, env.TM_CURRENT_LINE)
-	env.TM_LINE_INDEX = tostring(pos[1])
-	env.TM_LINE_NUMBER = tostring(pos[1] + 1)
-	env.SELECT_RAW, env.SELECT_DEDENT, env.TM_SELECTED_TEXT =
-		util.get_selection()
-end
+local builtin_vars = require("luasnip.util._builtin_vars")
 
-local lazy_vars = {}
+local function tbl_to_lazy_env(tbl)
+	local function wrapper(varname)
+		local val_ = tbl[varname]
+		if type(val_) == "function" then
+			return val_()
+		end
+		return val_
+	end
 
-local Environ = {}
-function Environ:new(pos, o)
-	o = o or {}
-	setmetatable(o, self)
-	_fill_eager_vars(o, pos)
-	return o
+	return wrapper
 end
 
-function Environ:__index(key)
-	local v = lazy_vars[key]()
-	rawset(self, key, v)
-	return v
-end
+local namespaces = {
+	[""] = {
+		init = builtin_vars.eager,
+		vars = tbl_to_lazy_env(builtin_vars.lazy),
+		eager = {},
+	},
+}
 
-function Environ:override(env, new_env)
-	for k, v in pairs(new_env) do
-		env[k] = v
+-- Namespaces allow users to define their own environmet variables
+local function _resolve_namespace_var(full_varname)
+	local parts = vim.split(full_varname, "_")
+	if #parts < 2 then
+		return nil
 	end
-end
+	local nmsp = namespaces[parts[1]]
 
--- Variables defined in https://code.visualstudio.com/docs/editor/userdefinedsnippets#_variables
-
--- Inherited from TextMate
-function lazy_vars.TM_FILENAME()
-	return vim.fn.expand("%:t")
-end
+	local varname
 
-function lazy_vars.TM_FILENAME_BASE()
-	return vim.fn.expand("%:t:s?\\.[^\\.]\\+$??")
-end
-
-function lazy_vars.TM_DIRECTORY()
-	return vim.fn.expand("%:p:h")
-end
-function lazy_vars.TM_FILEPATH()
-	return vim.fn.expand("%:p")
+	if nmsp then
+		varname = full_varname:sub(#parts[1] + 2)
+	else
+		nmsp = namespaces[""]
+		varname = full_varname
+	end
+	return nmsp.vars(varname)
 end
 
--- Vscode only
+local Environ = {}
 
-function lazy_vars.CLIPBOARD() -- The contents of your clipboard
-	return vim.fn.getreg('"', 1, true)
+-- returns nil, but that should be alright.
+-- If not, use metatable.
+function Environ.is_table(key)
+	return builtin_vars.is_table[key]
 end
 
-local function buf_to_ws_part()
-	local LSP_WORSKPACE_PARTS = "LSP_WORSKPACE_PARTS"
-	local ok, ws_parts = pcall(vim.api.nvim_buf_get_var, 0, LSP_WORSKPACE_PARTS)
-	if not ok then
-		local file_path = vim.fn.expand("%:p")
+function Environ:new(pos, o)
+	o = o or {}
+	setmetatable(o, self)
 
-		for _, ws in pairs(vim.lsp.buf.list_workspace_folders()) do
-			if file_path:find(ws, 1, true) == 1 then
-				ws_parts = { ws, file_path:sub(#ws + 2, -1) }
-				break
+	for ns_name, ns in pairs(namespaces) do
+		local eager_vars = {}
+		if ns.init then
+			eager_vars = ns.init(pos)
+		end
+		for _, eager in ipairs(ns.eager) do
+			if not eager_vars[eager] then
+				eager_vars[eager] = ns.vars(eager)
```

Comment:
Here it wouldn't make a difference (well, for our builtin vars at least) if the existing `eager_var` would be overwritten, correct? (don't necessarily want to change this, just to be sure)

Reply:
Yes, given that `init` and `eager` are potentially capable of providing the same vars and that they don't have to generate the same value, I had to decide to give priority to one of them. If you don't like that choose let me know, but it's one set or the other.

Reply:
Ah, right, I see now that this is just namespace-local.
This should work alright :+1: